httpgrpc/io.go

```diff
@@ -104,14 +104,52 @@ func asMetadata(header http.Header) (metadata.MD, error) {
 	return md, nil
 }
 
-func toHeaders(md metadata.MD, h http.Header) {
+var reservedHeaders = map[string]struct{}{
+	"accept-encoding":   {},
+	"connection":        {},
+	"content-type":      {},
+	"content-length":    {},
+	"keep-alive":        {},
+	"te":                {},
+	"trailer":           {},
+	"transfer-encoding": {},
+	"upgrade":           {},
+}
+
+func toHeaders(md metadata.MD, h http.Header, prefix string) {
 	// binary headers must be base-64-encoded
 	for k, vs := range md {
+		lowerK := strings.ToLower(k)
+		if _, ok := reservedHeaders[lowerK]; ok {
+			// ignore reserved header keys
+			continue
+		}
+		isBin := strings.HasSuffix(lowerK, "-bin")
 		for _, v := range vs {
-			if strings.HasSuffix(strings.ToLower(k), "-bin") {
+			if isBin {
 				v = base64.URLEncoding.EncodeToString([]byte(v))
 			}
-			h.Add(k, v)
+			h.Add(prefix+k, v)
 		}
 	}
 }
+
+type strAddr string
```

Comment:
This type and the impl below were just copied from the `google.golang.org/grpc/transport` package. So we create a peer address the same way as normal HTTP/2 gRPC when it is installed as an HTTP handler (vs. stand-alone server).